wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1194 @@
+    cur_csim_len += (len_csim_str[1] - '0') * 10;
+    if (len_csim_str[0] >= '0' &&
+            len_csim_str[0] <= '9') {
+        cur_csim_len += len_csim_str[0] * 100;
+    } else if (len_csim_str[0] != ' ') {
+        return BAD_FUNC_ARG;
+    }
+
+    if (cur_csim_len > IOTSAFE_CMDSIZE_MAX) {
+        return MEMORY_E;
+    }
+
+    /* Read out current Lc parameter in the CSIM command, last byte in the
+     * header
+     */
+    if (hex_to_bytes(cmd + AT_CSIM_CMD_SIZE + AT_CMD_LC_POS, &cur_lc, 1) < 0)
+        return BAD_FUNC_ARG;
+
+    /* Increase Lc and CSIM length according to the TLV len */
+    cur_lc += 1 + taglen_size + len;
+    cur_csim_len += 2 + (2 * taglen_size) + 2*len;
+
+    /* Position the out buffer to the current end of string */
+    out = cmd + XSTRLEN(cmd);
+    off = 0;
+
+    /* Write Tag */
+    bytes_to_hex(&tag, out, 1);
+    off += 2;
+
+    /* Write Len */
+    if (taglen_size == 2) {
+        bytes_to_hex(&len_b[1], out + 2, 1);
+        bytes_to_hex(&len_b[0], out + 4, 1);
+        off += 4;
+    } else {
+        bytes_to_hex(&len_b[0], out + 2, 1);
+        off += 2;
+    }
+
+    /* Write Val */
+    ret = bytes_to_hex(val, out + off, len);
+    off += ret;
+
+    /* Terminate string */
+    out[off] = 0;
+
+    /* Write new CSIM command size */
+    if (cur_csim_len > 99)
+        len_csim_str[0] = (cur_csim_len / 100) + '0';
+    else
+        len_csim_str[0] = ' ';
+    len_csim_str[1] = ((cur_csim_len % 100) / 10) + '0';
+    len_csim_str[2] = (cur_csim_len % 10) + '0';
+
+    /* Write new Lc value */
+    bytes_to_hex(&cur_lc, cmd + AT_CSIM_CMD_SIZE + AT_CMD_LC_POS, 1);
+    return off;
+}
+
+static int iotsafe_cmd_add_tlv16(char *cmd, byte tag, uint16_t len,
+        const byte *val)
+{
+    return iotsafe_cmd_add_tlv_ex(cmd, tag, len, val, 2);
+}
+
+static int iotsafe_cmd_add_tlv(char *cmd, byte tag, byte len, const byte *val)
+{
+    return iotsafe_cmd_add_tlv_ex(cmd, tag, len, val, 1);
+}
+
+static void iotsafe_cmd_complete(char *cmd)
+{
+    char *out = cmd + XSTRLEN(cmd);
+    out[0] = '"';
+    out[1] = '\r';
+    out[2] = '\n';
+    out[3] = 0;
+}
+
+/** IOT-Safe Commands
+ *
+ *
+ */
+static int expect_csim_response(const char *cmd, word32 size, char **reply)
+{
+    const char csim_response_hdr[] = "+CSIM:";
+    uint8_t len = 0;
+    char *csim_reply;
+    int ret;
+    char *payload;
+    char *endstr;
+    ret = expect_tok(cmd, size, csim_response_hdr, &csim_reply);
+    while (ret == 0) {
+        ret = expect_tok("AT\r\n", 4, csim_response_hdr, &csim_reply);
+    }
+    if (ret < 1)
+        return ret;
+    payload = XSTRSTR(csim_reply, "\"");
+    if (!payload)
+        return -1;
+    payload++;
+    if (XSTRNCMP(payload, "61", 2) == 0) {
+        if (hex_to_bytes(payload + 2, &len, 1) == 1) {
+            iotsafe_cmd_start(csim_cmd,1,IOTSAFE_INS_GETRESPONSE, 0, 0);
+            bytes_to_hex(&len, csim_cmd + AT_CSIM_CMD_SIZE + 8, 1);
```

Comment:
Would it be better to define `8` as well? Not to have a magic number :mage:. Something line `AT_CMD_HDR_LEN_OFFSET` (if I understand correctly).

Reply:
whoops, it was defined already as `AT_CMD_LC_POS`. Fixed.

Reply:
Still magical in `iotsafe_getrandom` on line 472 and iotsafe_cmd_start at line 200. Perhaps it would also be better to have the offsets in `iotsafe_cmd_start` be defined as well?